### ISSUE TITLE
Haskell Implementation of MicroQiskit

### ISSUE THREAD

**Issue by aditya-giri** [qiskit-advocate/qiskit-advocate-projects#11]
# Abstract

This is a Haskell implementation of James Wootton's MicroQiskit, aimed at providing a functional programming interface to MicroQiskit.

# Description

We have to start from scratch. We need to first complete the following steps:

- Implement the basic QuantumCircuit functions
- Implement the simulator
- Implement tests

Finally we have to figure out how to port the jobs to python-compatible code.

# Members
- @aditya-giri 
- @e-eight
- @spencerking 
- @kenixer

# Deliverable

A Haskell library implementing MicroQiskit.

# Github repo

Repo: [https://github.com/aditya-giri/MicroQiskit](https://github.com/aditya-giri/MicroQiskit)
Branch: [haskell](https://github.com/aditya-giri/MicroQiskit/tree/haskell)

**Comment by e-eight** [comment]
From Junye: project details are added to the issue.

**Comment by icepolarizer** [comment]
Brilliant! 🎉

**Comment by e-eight** [comment]
If you want to contribute, but are new to Haskell, here are a few resources to learn it "quickly".

http://yannesposito.com/Scratch/en/blog/Haskell-the-Hard-Way/
http://learnyouahaskell.com/
http://shuklan.com/haskell/index.html

**Comment by icepolarizer** [comment]
Nice. I used Haskell a bit...a long time ago, but I forgot most of them. I'll use these resources to reboot my memories.

**Comment by aditya-giri** [comment]
There's also [this book](https://haskellbook.com/) that I like. You'll find it on [libgen](http://libgen.rs/search.php?req=haskell+programming+from+first&lg_topic=libgen&open=0&view=simple&res=25&phrase=1&column=def).

**Comment by spencerking** [comment]
I'd like to help out as well. I think this will require a bit more upfront planning than the other implementations since it's a completely different paradigm.

**Comment by HuangJunye** [comment]
I just got to know about Haskell when I started learning Rust a few days ago to work on project #4 . The only thing I know about Haskell is that it's not object oriented language but a functional language. But there is already a C implementation of MicroQiskit so I assume it won't be that different.

**Comment by e-eight** [comment]
> There's also [this book](https://haskellbook.com/) that I like. You'll find it on [libgen](http://libgen.rs/search.php?req=haskell+programming+from+first&lg_topic=libgen&open=0&view=simple&res=25&phrase=1&column=def).

That’s a great book too! And they keep on updating it.

**Comment by e-eight** [comment]
> I just got to know about Haskell when I started learning Rust a few days ago to work on project #4 . The only thing I know about Haskell is that it's not object oriented language but a functional language. But there is already a C implementation of MicroQiskit so I assume it won't be that different.

Functional programming is very different from imperative programming. In the C/C++ ecosphere only thing that would be similar to a functional programming language is template metaprogramming in C++. That being said, it is possible to write imperative style code in Haskell, so the end result could be similar.

**Comment by aditya-giri** [comment]
A lot of languages including Python are functional too, just not purely functional like Haskell is. Imperative style code in Haskell is generally frowned upon but you do see some use-cases occasionally. I think this will be good fun! :D

**Comment by kenixer** [comment]
I was thinking about Racket implementation at first, but quickly changed my mind. Haskell's architecture and the syntax look great on Qiskit. Infix function can be advantageous for this kind of operation. If we can complete this, it's like we're creating the most intuitive MicroQiskit implementation!

**Comment by aditya-giri** [comment]
What do you guys think is the best way to get familiar enough with Haskell? Reading a book like the one I linked above could take ages, but `learnyouhaskell` is probably the most convenient for now?

**Comment by e-eight** [comment]
Yes that is over 1000 pages in its latest incarnation. You could follow this. https://github.com/bitemyapp/learnhaskell (just the first course) or this https://crypto.stanford.edu/~blynn/haskell/. Followed by this https://en.wikibooks.org/wiki/Write_Yourself_a_Scheme_in_48_Hours. I am planning to do the last one over a weekend or something. That should be enough to get started, and then we learn as we progress in the project.

**Comment by icepolarizer** [comment]
For now, I think _Haskell the Hard Way_ seems to work best for me. Quick and dense.

**Comment by kenixer** [comment]
By the way, I think the _Members_ section should be updated. We have four assignees, but there's only two in the list.

**Comment by icepolarizer** [comment]
@HuangJunye By the way, I'd like to suggest updating the Members section in project details. We have four assignees, but there's only two in the list.

**Comment by aditya-giri** [comment]
One last resource share: https://www.youtube.com/channel/UC3xdLFFsqG701QAyGJIPT1g

**Comment by aditya-giri** [comment]
Can we decide on a date to regroup after getting a little familiar with Haskell?

**Comment by e-eight** [comment]
Does a week sound fair?

**Comment by aditya-giri** [comment]
I'm gonna need a little longer, I'd prefer 2 weeks. What about the others?

**Comment by icepolarizer** [comment]
Two weeks seems legit for me :)

**Comment by icepolarizer** [comment]
Also, I think the conversion of OOP -> FP will be the most challenging part for us. To be honest, it seems impossible to completely copy the architecture of MicroQiskit, because it's completely OOP based. We have to think more in-depth about imitating them.

**Comment by aditya-giri** [comment]
You're right, we will have to give it some thought. It's alright if it takes time, as long as we do a good job...
(Also, we can set up a private channel on slack once we start so we don't have to discuss everything here)

**Comment by icepolarizer** [comment]
Agreed! I'll make one.
And I think the name should be **'functional-microqiskit'**, because it might be a good idea to make Qiskit for other FP languages once we complete the theorem of FP-styled Qiskit.

**Comment by icepolarizer** [comment]
Looks like I can't make another channel in Qiskit slack. @HuangJunye could you please help?

**Comment by HuangJunye** [comment]
@MonolithPenguin Done! Love the enthusiasm of this team!
